bundle/libraries/match.go

```diff
@@ -2,44 +2,79 @@ package libraries
 
 import (
 	"context"
+	"fmt"
+	"path/filepath"
 
 	"github.com/databricks/cli/bundle"
 	"github.com/databricks/cli/libs/diag"
+	"github.com/databricks/databricks-sdk-go/service/compute"
 	"github.com/databricks/databricks-sdk-go/service/jobs"
 )
 
 type match struct {
 }
 
-func MatchWithArtifacts() bundle.Mutator {
+func ValidateLocalLibrariesExist() bundle.Mutator {
 	return &match{}
 }
 
 func (a *match) Name() string {
-	return "libraries.MatchWithArtifacts"
+	return "libraries.ValidateLocalLibrariesExist"
 }
 
 func (a *match) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
-	tasks := findAllTasks(b)
-	for _, task := range tasks {
-		if isMissingRequiredLibraries(task) {
-			return diag.Errorf("task '%s' is missing required libraries. Please include your package code in task libraries block", task.TaskKey)
+	for jobKey, job := range b.Config.Resources.Jobs {
+		err := validateEnvironments(job.Environments, b, jobKey)
+		if err != nil {
+			return diag.FromErr(err)
 		}
-		for j := range task.Libraries {
-			lib := &task.Libraries[j]
-			_, err := findArtifactFiles(ctx, lib, b)
+
+		for i, task := range job.JobSettings.Tasks {
+			err := validateTaskLibraries(task.Libraries, b, jobKey, i)
 			if err != nil {
 				return diag.FromErr(err)
 			}
 		}
 	}
+
+	return nil
+}
+
+func validateTaskLibraries(libs []compute.Library, b *bundle.Bundle, jobKey string, taskIndex int) error {
+	for i, lib := range libs {
+		path := libraryPath(&lib)
+		if path == "" || !IsLocalPath(path) {
+			continue
+		}
+
+		loc := b.Config.GetLocation(fmt.Sprintf("resources.jobs.%s.tasks.%d.libraries.%d", jobKey, taskIndex, i))
+		matches, err := findMatches(filepath.Join(filepath.Dir(loc.File), path), b)
+		if err != nil {
+			return err
+		}
+
+		if len(matches) == 0 {
+			return fmt.Errorf("file %s is referenced in libraries section but doesn't exist on the local file system", libraryPath(&lib))
+		}
+	}
+
 	return nil
 }
 
-func isMissingRequiredLibraries(task *jobs.Task) bool {
-	if task.Libraries != nil {
-		return false
+func validateEnvironments(envs []jobs.JobEnvironment, b *bundle.Bundle, jobKey string) error {
+	for i, env := range envs {
+		for j, dep := range env.Spec.Dependencies {
+			loc := b.Config.GetLocation(fmt.Sprintf("resources.jobs.%s.environments.%d.spec.dependencies.%d", jobKey, i, j))
+			matches, err := findMatches(filepath.Join(filepath.Dir(loc.File), dep), b)
+			if err != nil {
+				return err
+			}
+
+			if len(matches) == 0 && IsLocalPath(dep) {
+				return fmt.Errorf("file %s is referenced in environments section but doesn't exist on the local file system", dep)
```

Comment:
We should test this with all valid options for dependencies to verify it only triggers for actual paths.

This might imply that we have to require a `./` prefix for local file references, or require a file extension.

Otherwise "somedep" as a PyPI reference cannot be distinguished from "somedep" as a local file.